### PR TITLE
Color conversion fix v6

### PR DESCRIPTION
**Pull request by akallabeth**
Superseeds #3450 (rebased to current master)

* Implements new colour format handling (NOTE: External Client/Server implementations must be adjusted)
* Client channels use Thread Local Storage for contexts now to allow multiple instances per process
* Removed internal functions from API headers, now in private headers
* Applied proper function scoping (static/library local/exported)
* Adjusted clients to new API, fixed some issues that were found during tests
* Deduplicated GFX channel code, adjusted callback API to reduce client side code
* Refactored GDI functions, now implementing all ROP operations
* Deactivated ```glyph-cache``` by default as it produces artefacts and therefore is not ready for general use
* Unified / made consistend API function arguments and return values
* Updated helper scripts for iOS and android
* Reformatted code touched with astyle helper script

### CODE REVIEW

**Comment by realjiangms** [comment]
There's too many file. I can't open the file changed tab(due to poor network) to add comments. So just paste in general comments. 

1. In rfx_process_message, last argument REGION16* invalidRegion
I guess it is an 'out' parameter that should be filled with invalid Region. Then it should be initialized if it is not NULL before we reach:
1194                     region16_union_rect(invalidRegion, invalidRegion, &updateRects[j]);
Or it might have following funny content:
(gdb) p *invalidRegion
$16 = {extents = {left = 29872, top = 60421, right = 32767, bottom = 0}, data = 0x7fffec0575b0}

2. Shadow server with basic codec (non-rfx, non-gfxh264) have wrong color format. I think the input color format need to be fixed as the format enum might have been redefined in this PR.
3. Shadow server with rfx shows mono picture. I hope it's same issue as $2
4. Shadow server with gfx doesn't accessible. Win10 mstsc just disconnect when it start to send out h264 stream.  The xfreerdp from this PR also fails with gfx-h264 as following:
[00:43:20:352] [23651:23662] [ERROR][com.freerdp.channels.rdpgfx.client] - rdpgfx_recv_wire_to_surface_1_pdu failed with error 13!
[00:43:20:352] [23651:23662] [ERROR][com.freerdp.channels.rdpgfx.client] - Error while parsing GFX cmdId: RDPGFX_CMDID_WIRETOSURFACE_1 (0x0001)
[00:43:20:352] [23651:23662] [ERROR][com.freerdp.channels.rdpgfx.client] - rdpgfx_recv_pdu failed with error 13!
[00:43:20:352] [23651:23662] [ERROR][com.freerdp.channels.drdynvc.client] - drdynvc_order_recv failed with error 13!

But it's interesting that the xfreerdp from standard FreeRDP works fine, with just same color format issue as $2

**Comment by akallabeth** [comment]
@realjiangms Thx for the review.
As for your comments:

1. Well, that is not true. It is a out parameter but is initialized at the beginning of a decoding block. (See ```gdi_OutputUpdate```)
2. 3. 4. Ok, changed the color format. Does it work for you now?

File: channels/drdynvc/client/drdynvc_main.c [hunk 1]
```diff
@@ -1612,46 +1501,34 @@ BOOL VCAPITYPE VirtualChannelEntry(PCHANNEL_ENTRY_POINTS pEntryPoints)
...
+	        &drdynvc->channelDef, 1, VIRTUAL_CHANNEL_VERSION_WIN2000,
+	        drdynvc_virtual_channel_init_event);
 
 	if (CHANNEL_RC_OK != rc)
 	{
 		WLog_ERR(TAG, "pVirtualChannelInit failed with %s [%08X]",
-				 WTSErrorToString(rc), rc);
-		free(drdynvc);
-		free(*(pEntryPointsEx->ppInterface));
-		*(pEntryPointsEx->ppInterface) = NULL;
```
Comment (MartinHaimberger) [inline channels/drdynvc/client/drdynvc_main.c]: setting the ppInterface to NULL after freeing is gone missing .

Comment (MartinHaimberger) [inline channels/drdynvc/client/drdynvc_main.c]: This applies to all channels, please review all VirtualChannelEntryPoints where FALSE is returned.

File: channels/encomsp/client/encomsp_main.c [hunk 2]
```diff
@@ -1346,40 +1269,40 @@ BOOL VCAPITYPE VirtualChannelEntry(PCHANNEL_ENTRY_POINTS pEntryPoints)
...
+	encomsp->channelEntryPoints.ppInterface = &
+	        (encomsp->channelEntryPoints.pInterface);
+	s_TLSPluginContext = encomsp;
 	return TRUE;
 error_out:
+
 	if (isFreerdp)
 		free(encomsp->context);
+
 	free(encomsp);
```
Comment (MartinHaimberger) [inline channels/encomsp/client/encomsp_main.c]: *(pEntryPointsEx->ppInterface) = NULL;

File: channels/rail/client/rail_main.c [hunk 3]
```diff
@@ -874,42 +812,36 @@ BOOL VCAPITYPE VirtualChannelEntry(PCHANNEL_ENTRY_POINTS pEntryPoints)
...
-		WLog_ERR(TAG, "rail_add_init_handle_data failed with error %lu!", error);
-		goto error_out;
-	}
-
+	s_TLSPluginContext = rail;
 	return TRUE;
 error_out:
+
 	if (isFreerdp)
 		free(rail->context);
```
Comment (MartinHaimberger) [inline channels/rail/client/rail_main.c]: *(pEntryPointsEx->ppInterface) = NULL;

File: client/Mac/MRDPView.m [hunk 4]
```diff
@@ -182,54 +179,36 @@ DWORD mac_client_thread(void* param)
...
-		else
-		{
-			if (!(updateEvent = freerdp_get_message_queue_event_handle(instance, FREERDP_UPDATE_MESSAGE_QUEUE)))
-			{
-				WLog_ERR(TAG, "failed to get update event handle");
-				goto disconnect;
-			}
-			events[nCount++] = updateEvent;
-		}
-
```
Comment (MartinHaimberger) [inline client/Mac/MRDPView.m]: Async Update removed from the mac client with purpose ?

File: client/Mac/MRDPView.m [hunk 5]
```diff
@@ -960,99 +917,80 @@ BOOL mac_pre_connect(freerdp* instance)
...
 	PubSub_SubscribeChannelDisconnected(instance->context->pubSub,
-					    (pChannelDisconnectedEventHandler) mac_OnChannelDisconnectedEventHandler);
+	                                    (pChannelDisconnectedEventHandler) mac_OnChannelDisconnectedEventHandler);
 
-	if (!freerdp_client_load_addins(instance->context->channels, instance->settings))
+	if (!freerdp_client_load_addins(instance->context->channels,
+	                                instance->settings))
 		return FALSE;
 
-	if (freerdp_channels_pre_connect(instance->context->channels, instance) != CHANNEL_RC_OK)
```
Comment (MartinHaimberger) [inline client/Mac/MRDPView.m]: removed by purpose ?

Comment (akallabeth) [inline client/Mac/MRDPView.m]: @MartinHaimberger Yes, moved to the core library (as every client needs to call that)

File: client/Mac/MRDPView.m (hunk 4 above)
Comment (akallabeth) [inline client/Mac/MRDPView.m]: @MartinHaimberger Yes, moved to the core library (as every client needs to call that)

**Comment by MartinHaimberger** [comment]
+1 LGTM as far as i can say

**Comment by akallabeth** [comment]
@freerdp-bot Don't be a coward, test this whole thing please!

**Comment by realjiangms** [comment]
I browsered the changes and I think they should definitely fix the color confusing. But I'm not sure whether they fix the h264 gfx issue by just code browsing. I'll update the code and test later

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by realjiangms** [comment]
@akallabeth

1. gfx with openh264 still not work. I'll try to figure it out.

2. For the issue I mentioned for last argument REGION16* invalidRegion in rfx_process_message, at least it is not initialized when it is called in xf_gdi_surface_bits

3. Any plan to re-enable optimized ssse3_YUV420ToRGB_8u_P3AC4R? Actaully the signature of this function (and general_YUV420ToRGB_8u_P3AC4R) don't need to be changed. The function name determined that they only apply to NV12=>BGRX conversion. The format of the function name should be derived from Intel IIP library

**Comment by realjiangms** [comment]
Resolved $1 and $2 in privious note.
1. Well, the definition of RDPGFX_SURFACE_COMMAND.format is changed. I'll submit an PR to decode it before writing pixelFormat.

2. For  rfx_process_message called in xf_gdi_surface_bits, it seems the invalidRegion is never used. We might need to send NULL (same as gdi_surface_bits)

2.

**Pull request merged**